File: Labs/Lab 7/C#/Exp/Expression.cs
using System;

public class Expression
{
    public double A { get; set; }
    public double B { get; set; }
    public double C { get; set; }
    public double D { get; set; }

    public Expression(double a, double b, double c, double d)
    {
        A = a; B = b; C = c; D = d;
    }

    public double Calculate()
    {
        double numerator = (4 * B - C) * A;
        double denominator = B + C / D - 1;

        if (numerator <= 0)
            throw new ArgumentException("Логарифм від <= 0");

        if (denominator == 0)
            throw new DivideByZeroException("Ділення на нуль");

        return Math.Log10(numerator) / denominator;
    }

    public override string ToString()
    {
        return $"a={A} b={B} c={C} d={D}";
    }
}

File: Labs/Lab 7/C++/Exp/Expression.cpp
#include "Expression.h"
#include <cmath>
#include <stdexcept>
#include <sstream>

Expression::Expression(double a, double b, double c, double d) : a(a), b(b), c(c), d(d) {}

double Expression::calculate() const {
    double numerator = (4 * b - c) * a;
    double denominator = b + c / d - 1;

    if (numerator <= 0) throw std::domain_error("Логарифм від <= 0");
    if (denominator == 0) throw std::runtime_error("Ділення на нуль");

    return log10(numerator) / denominator;
}

std::string Expression::toString() const {
    std::ostringstream oss;
    oss << "a=" << a << " b=" << b << " c=" << c << " d=" << d;
    return oss.str();
}
